File: dart/samples/pond/editors.js
// Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

var ERROR_KIND = 1;
var WARNING_KIND = 2;
var ERROR_CLASSNAME = "compile_error";
var WARNING_CLASSNAME = "compile_warning";

var editors = {};
var markers = [];

function newEditor(id, type, listener) {
  editors[id] = CodeMirror(document.getElementById(id), {
    mode: type,
    tabSize: 2,
    lineNumbers: true,
    gutter: true,
    onChange: listener
  });
}

function changeListener(id) {
  return function(editor, textChanges) {
    window.postMessage(['js-to-dart', 'update', id], '*');
  }
}

function newMark(editorId, startLine, startCol, endLine, endCol, kind) {
  var className = "";
  if (kind == ERROR_KIND) {
    className = ERROR_CLASSNAME;
  } else if (kind == WARNING_KIND) {
    className = WARNING_CLASSNAME;
  }
  var marker = editors[editorId].markText(
      {line: startLine, ch: startCol}, {line: endLine, ch: endCol}, className);
  var markerId = markers.length;
  markers.push(marker);
  return markerId;
}

function messageDispatcher(envelope) {
  if (envelope[0] != 'dart-to-js') return;
  var returnId = envelope[1];
  var message = envelope[2];
  var command = message[0];
  var args = message[1];
  var reply = null;
  switch (command) {
    case "newEditor":
      newEditor(args[0], args[1], (args[2] ? changeListener(args[0]): null));
      break;
    case "getText":
      reply = editors[args[0]].getValue();
      break;
    case "setText":
      editors[args[0]].setValue(args[1]);
      break;
    case "refresh":
      editors[args[0]].showLine(0);
      break;
    case "mark":
      var reply = newMark(
          args[0], // editor
          args[1], args[2], // start line & column
          args[3], args[4], // end line & column
          args[5]); // kind
      break;
    case "clearMark":
      markers[args[0]].clear();
      break;
  }
  window.postMessage(['js-to-dart-reply', returnId, reply], '*');
}

window.addEventListener('message',
    function (e) { messageDispatcher(e.data); }, false);

File: dart/samples/pond/editors.dart
// Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

// A thin Dart-API to communicate with an editor service that is running in a
// separate isolate.
#library("editors");

/** An editor factory that creates new editor instances. */
interface EditorFactory {

  /**
   * Create an editor of a specific [type] (dart, js, html) and display it under
   * a DOM element with the given [id].
   */
  Future<Editor> newEditor(String id, String type, [Function changeListener]);
}

/** A remote-editor interface. */
interface Editor {

  /** Asynchronously retrieve the contents of the editor. */
  Future<String> getText();

  /**
   * Asynchronously update the contents of the editor. The returned future will
   * be completed when the text has been updated.
   */
  Future setText(String value);

  /** Create an error or warning marker between [start] and [end]. */
  Future<Marker> mark(Position start, Position end, int kind);

  /**
   * Ensure that an editor is visible and up to date. Sometimes editors are not
   * up to date if the UI is hidden when it was rendered.
   */
  Future refresh();
}

/** Interface for a text-marker in an editor. */
interface Marker {
  Future clear();
}

/** Enumeration of kinds of markers. */
class Marks {
  static final NONE = 0;
  static final ERROR = 1;
  static final WARNING = 2;
  static final INFO = 3;
}

/** Represents a position in the editor. */
class Position {
  int line;
  int column;
  Position(this.line, this.column);
}
